rfcs/text/0000-prefab-structs.md

```diff
@@ -0,0 +1,128 @@
+- Feature Name: prefab-structs
+- Start Date: 2018-04-24
+- RFC PR: (leave this empty)
+- Feature Commit(s): (leave this empty)
+
+# Summary
+
+Typed Racket's handling of #:prefab structs is currently (as of April 2018) unsound.
+Instead of treating them like standard structs where we can assume (and enforce)
+that fields always contain values of a certain type (i.e. the current behabior), 
+prefabs should more-or-less be treated exactly like tagged n-ary cons/mcons cells.
+
+# Motivation
+
+Prefabs are clearly useful and desirable (e.g. they're serializable) and our current
+lack of sound support for them should be addressed.
+
+# Guide-level explanation
+
+## Prefab Top Types
+
+`(PrefabTop name)` is the top type for the prefab described by the
+identifier `name`. For immutable prefab structs, this type is equivalent
+(in fact short hand for) the type `(Prefab name Any ...)` (where there are
+as many occurrences of `Any` as there are fields in the prefab struct
+described by `name`). For mutable prefab structs, this type can be used for
+field reads (i.e. always producing a value of type `Any`) but cannot be used 
+for field writes (since the type of the invariant field is not known).
+
+
+## Immutable Prefabs
+
+This __immutable prefab defininition__:
+
+```racket
+(struct point ([x : Integer] [y : Integer])
+  #:prefab)
+```
+
+defines types/constructors/accessors for what can be thought of as a
+ _named tuple_ type, producing (more-or-less) the following type definition:
+
+`point ≈ (Prefab point Integer Integer)`
+
+and the following functions:
+
+- `point` of type `(-> Integer Integer point)`
+
+- `point?` of type `(-> Any Boolean : (Prefab point Any Any))` 
+(_NOTE: the type has `Any` not `Integer` in the field! This is because
+        a prefab could have been created anywhere with any value in
+        its fields_)
```

Comment:
And `(make-predicate point)` will have type `(-> Any Boolean : point)`?

Reply:
I would think so -- the generated predicate would of course need to inspect the fields. I'll add a note to the RFC.